### ISSUE TITLE
Closed Anonymous Modules

### ISSUE THREAD

**Issue by larrytheliquid** [agda/agda#2293]
I find the syntax for sections / anonymous modules (introduced in #735) to be unnatural. 

Once you close a definition of a named module, there is no way to reopen it by adding more definitions with the old ones in scope. This is fine, and you end up getting a sensible duplicate name error if you try:

```agda
module Foo where
  postulate Collection : (A : Set) -> Set

module Foo where
  postulate empty : {A : Set} -> Collection A
-- Duplicate definition of module Foo
```

However, if you do the same thing with anonymous modules it almost feels like there is a single anonymous module that you can "reopen" to extend at any time.

```agda
module _ where
  postulate Collection : (A : Set) -> Set

module _ where
  postulate empty : {A : Set} -> Collection A
```

This is because an anonymous module desugars into a private module with a fresh name, that is automatically opened. I propose changing the syntax of anonymous modules that are automatically opened (sections) to use "..." as their name rather than underscore. This would allow anonymous modules named "_" to have the same desugaring, but without opening the module.

Sections with same old semantics, just renaming "_" to "...".
```agda
module ... params where
  declarations
-- desugars to:
private
  module Fresh params where
    declarations
open Fresh public
```

Anonymous modules as sections that are not automatically opened:
```agda
module _ params where
  declarations
-- desugars to:
private
  module Fresh params where
    declarations
```

Now we can still write the former example, but it feels more natural syntactically.
```agda
module ... where
  postulate Collection : (A : Set) -> Set

module ... where
  postulate empty : {A : Set} -> Collection A
```

But we can also write examples using the old syntax, where each anonymous module does not have previous definitions in scope:
```agda
module _ where
  data List (A : Set) : Set

module _ where
  data List (A : Set) : Set
-- does NOT raise the error "Multiple definitions of List"
```

Using underscore for anonymous modules that are never opened is useful for documentation, where you write a version of datatypes/functions that is inefficient or not structurally terminating as documentation in the code that is never exported. Then, you can define the datatype with the same name outside of the anonymous module with the proper implementation without getting a name collision. I do this a lot when writing literate agda documents, where I showcase many different ways to implement functions but use the same names for each implementation (each one in a different module).

The unopened semantics for underscore also makes more sense syntactically with the new anonymous declarations feature (63fe72ad7a4462d05517a65f7dbcc371cfe69b9d). Specifically, both features use underscore to designate writing definitions that do not pollute the scope. 

Note that this is merely an argument for syntactic changes to make writing Agda feel more natural and make its features feel more cohesive. Currently, I write the previous example as follows:
```agda
module _ where
  private
    data List (A : Set) : Set

module _ where
  private
    data List (A : Set) : Set
-- does NOT raise the error "Multiple definitions of List"
```

This works fine, but the multiple levels of indentation are a bit annoying, and as explained above I feel like the recommended syntactic changes make more sense. This change is easy to implement, and I would be happy to do it myself, I just would like to know if such a change would be merged or not (I realize it would break old code, hence why I'm asking.)

**Comment by larrytheliquid** [comment]
Also, here is a self-contained example of why we might want both open and closed sections. The example shows how we might define the partial `head` function using a `Maybe` return type in a non-dependent language, then shows how to define a similar function taking advantage of dependent types. The one we want in scope is the dependently typed version (so we use an open section via ellipses), but it is nice to be able to keep the non-dependent version as an example (so we use a closed section via underscore).

``` agda
record ⊤ : Set where
  constructor tt

data Maybe (A : Set) : Set where
  nothing : Maybe A
  just : A → Maybe A

data List (A : Set) : Set where
  nil : List A
  cons : A → List A → List A

module _ {A : Set} where
  Head : Set
  Head = Maybe A

  head : List A → Head
  head nil = nothing
  head (cons x xs) = just x

module ... {A : Set} where
  Head : List A → Set
  Head nil = ⊤
  Head (cons x xs) = A

  head : (xs : List A) → Head xs
  head nil = tt
  head (cons x xs) = x
```

**Comment by UlfNorell** [comment]
I don't agree that using `...` instead of `_` for the module name would be more natural. Currently we use `...` in two places:
- with-clauses that repeat their parent patterns, and
- instance open (`open M {{...}}`)

In both cases the meaning is (sort of) "the obvious thing", which doesn't quite match the use as an anonymous module name.

Using `_` makes more sense to me, and I guess the only thing you find unnatural is that anonymous modules are automatically opened. I would argue that opening is the natural thing to do since otherwise the module is useless except as documentation. Even though documentation is important, I don't think we should optimise the language for it. Making you explicitly write `private` in this case seems quite reasonable (and you can always indent the `private` just a single space to avoid excess indentation).

I definitely don't think it's worth breaking everybody's code over this.

**Comment by andreasabel** [comment]
The issue here seems that there should be a way to define throw-away modules just like there are throw-away declarations now (which I was not aware of until now).
Maybe throw-away declarations and modules should both not use a single underscore `_`.  Instead, they could use a double underscore `__`. That would be consistent. Re #1465.

**Comment by gallais** [comment]
`_` seems to be the usual token for "please throw the binding away" in patterns too (and by extension in `let`s):

```
A : Set₁
A = let B = Set
        _ = B
    in B
```

`...` is not necessarily the appropriate name for section-like modules but I do like the idea of having `_` mean across the board that the corresponding bindings are thrown away.

**Comment by andreasabel** [comment]
Maybe `module _` for sections is ugly anyway (so is `module ...`) and we should add a keyword `section` instead.
Still this would not make `module _` available for throw-away modules for a while (because we need a deprecation phase).

**Comment by larrytheliquid** [comment]
I was thinking this way `...` gets consistently used to add something while implicitly using the old version of something else:
- with-clauses that repeat their parent patterns
  - add new function clause, using old patterns
- instance record modules (`M {{...}}`)
  - add new instance-arg record, using old definitions
- automatically opened sections
  - add new section definitions, using old scope

But I do like the idea of using a new `section` keyword instead, since sections are what we are really talking about. Maybe something like:

```
section private {A : Set} where
  Head : Set
  Head = Maybe A

  head : List A → Head
  head nil = nothing
  head (cons x xs) = just x

section public {A : Set} where
  Head : List A → Set
  Head nil = ⊤
  Head (cons x xs) = A

  head : (xs : List A) → Head xs
  head nil = tt
  head (cons x xs) = x
```

We could also make `section` by default desugar to `section public`?

**Comment by UlfNorell** [comment]
I could get behind a new `section` keyword. It looks nice and would let us keep `module _` for backwards compatibility.

**Comment by fredrikNordvallForsberg** [comment]
How about `private section` instead of `section private`? And `section` on its own should mean the current automatically opened `module _` right?

**Comment by larrytheliquid** [comment]
Yeah, I think that would make sense:

```
private section {A : Set} where
  Head : Set
  Head = Maybe A

  head : List A → Head
  head nil = nothing
  head (cons x xs) = just x

-- currently module _ {A : Set} where
section {A : Set} where
  Head : List A → Set
  Head nil = ⊤
  Head (cons x xs) = A

  head : (xs : List A) → Head xs
  head nil = tt
  head (cons x xs) = x
```

_edit_
Ah nevermind, it looks like there would be parsing ambiguity because currently this is allowed:

```
private module Foo where
```

Although I'm not sure what it means, since I get a parse error when trying to add anything indented underneath?

_edit_
I see now, you have to indent really far when using that syntax:

```
private data Foo : Set where
         foo : Foo
```

Also, it doesn't actually make `Foo` private... I kind of understand why but it's confusing.

This ambiguity would not be there for `section private args where` though.

**Comment by andreasabel** [comment]
If #1145 was fixed then both

```
module _ tel where private
  ds
private module _ tel where
  ds
```

would parse.  The first alternative does already the right thing!?  Except maybe that you can do

```
module _ where
  private
    module M where
       postulate A : Set
open M public
```

I.e. the scope of private does not extend into submodules.
Is the main concern in this issue that

```
module  _ where private
  ds
```

does not parse?

**Comment by larrytheliquid** [comment]
Good catch about the `private module _` causing submodules to still get into scope because the anonymous module is opened using `public`.

```
private
  module _ where
    module Foo where
-- Foo in scope

module _ where
  private
    module Bar where
-- Bar not in scope
```

In that case, yes I think having the `module _ where private` version parse would be what I'd like.

**Comment by andreasabel** [comment]
I am marking this as blocked by #1145.

**Comment by gallais** [comment]
Something worth noting following @txa's message to the mailing list:
the `private section` approach would a priori not allow you to repeat
yourself for pedagogical purposes because all of the `private` blocks
in a file live in the same `private` scope.

**Comment by anuyts** [comment]
I disagree that it is fine that we cannot extend existing modules within the same file (see #6038).

Also I don't think `section` is necessarily the best name: nameless modules are for grouping content under a common telescope and do not necessarily correspond to sections or reasoning stages in a proof. I think `using (tele : Tele) in` or `abstract (tele : Tele) in` would have been a good syntax had the keywords `using` and `abstract` not already been in use for conceptually conflicting purposes. Perhaps `assume/given/let/with (tele : Tele) in/where/let/do`? Or `open module _ (tele : Tele) where` as I proposed in #6036, although this is of course more verbose.

**Comment by gallais** [comment]
FTR Idris uses `parameters`.

**Comment by anuyts** [comment]
Or, since (as mentioned in #6036) `open` is really a statement to define a nameless module by equality; defining a nameless module by providing its contents should perhaps also start with an `open` keyword. So if/when a syntax has been decided for `open` with telescope abstraction (as suggested in #6036), then we could write something like `open λ (tele : Tele) -> where`.
